File: backend/core/src/modules/media/media.service.ts
import { BadRequestException, ForbiddenException, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import crypto from 'crypto';
import fs from 'fs/promises';
import path from 'path';
import { Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { MediaAsset } from '../../entities';

@Injectable()
export class MediaService {
  constructor(
    private readonly projectAccessService: ProjectAccessService,
    @InjectRepository(MediaAsset)
    private readonly mediaRepo: Repository<MediaAsset>,
  ) {}

  private uploadDir(): string {
    return process.env.MEDIA_UPLOAD_DIR || '/app/uploads';
  }

  private async ensureUploadDir() {
    await fs.mkdir(this.uploadDir(), { recursive: true });
  }

  private extFromMime(mimeType: string): string {
    switch (mimeType) {
      case 'image/png':
        return '.png';
      case 'image/jpeg':
        return '.jpg';
      case 'image/webp':
        return '.webp';
      case 'image/gif':
        return '.gif';
      case 'image/svg+xml':
        return '.svg';
      default:
        return '';
    }
  }

  async upload(user: { userId?: string; role?: string } | undefined, file: Express.Multer.File | undefined, projectId?: string) {
    const userId = user?.userId;
    if (!userId) throw new ForbiddenException('로그인이 필요합니다.');
    if (!file) throw new BadRequestException('file이 필요합니다.');

    if (projectId) {
      await this.projectAccessService.assertProjectAccess(userId, projectId);
    }

    await this.ensureUploadDir();

    const id = crypto.randomUUID();
    const ext = this.extFromMime(file.mimetype) || path.extname(file.originalname || '') || '';
    const filename = `${id}${ext}`;
    const storagePath = path.join(this.uploadDir(), filename);
    const url = `/api/media/${id}`;

    await fs.writeFile(storagePath, file.buffer);

    const asset = await this.mediaRepo.save(
      this.mediaRepo.create({
        id,
        userId,
        projectId: projectId || undefined,
        originalName: file.originalname,
        mimeType: file.mimetype,
        size: file.size,
        storagePath,
        url,
      }),
    );

    return asset;
  }

  async findById(id: string): Promise<MediaAsset | null> {
    return this.mediaRepo.findOne({ where: { id } as any });
  }

  async assertOwnership(user: { userId?: string; role?: string } | undefined, asset: MediaAsset) {
    if (!user?.userId) throw new ForbiddenException('로그인이 필요합니다.');
    if (user.role === 'admin') return;
    if (asset.userId !== user.userId) {
      throw new ForbiddenException('권한이 없습니다.');
    }
  }

  async delete(user: { userId?: string; role?: string } | undefined, id: string) {
    const asset = await this.findById(id);
    if (!asset) throw new NotFoundException('파일을 찾을 수 없습니다.');
    await this.assertOwnership(user, asset);

    await this.mediaRepo.delete({ id } as any);
    try {
      await fs.unlink(asset.storagePath);
    } catch {
      // ignore
    }
    return { success: true };
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/MediaService.java
package store.pjcloud.cowrite.core.service;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.web.multipart.MultipartFile;
import store.pjcloud.cowrite.core.entity.MediaAsset;
import store.pjcloud.cowrite.core.repository.MediaAssetRepository;

@Service
public class MediaService {
    private final ProjectAccessService projectAccessService;
    private final MediaAssetRepository mediaRepo;

    public MediaService(ProjectAccessService projectAccessService, MediaAssetRepository mediaRepo) {
        this.projectAccessService = projectAccessService;
        this.mediaRepo = mediaRepo;
    }

    private String uploadDir() {
        String dir = System.getenv("MEDIA_UPLOAD_DIR");
        return dir == null || dir.isBlank() ? "/app/uploads" : dir;
    }

    private String extFromMime(String mimeType) {
        if (mimeType == null) return "";
        return switch (mimeType) {
            case "image/png" -> ".png";
            case "image/jpeg" -> ".jpg";
            case "image/webp" -> ".webp";
            case "image/gif" -> ".gif";
            case "image/svg+xml" -> ".svg";
            default -> "";
        };
    }

    public MediaAsset upload(AiService.UserContext user, MultipartFile file, String projectId) {
        if (user == null || user.userId() == null) throw new ResponseStatusException(HttpStatus.FORBIDDEN, "로그인이 필요합니다.");
        if (file == null || file.isEmpty()) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "file이 필요합니다.");

        UUID projectUuid = null;
        if (projectId != null && !projectId.isBlank()) {
            projectUuid = UUID.fromString(projectId);
            projectAccessService.assertProjectAccess(user.userId(), projectUuid);
        }

        try {
            Files.createDirectories(Paths.get(uploadDir()));
        } catch (IOException ex) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "업로드 폴더를 생성할 수 없습니다.");
        }

        UUID id = UUID.randomUUID();
        String ext = extFromMime(file.getContentType());
        if (ext.isBlank() && file.getOriginalFilename() != null) {
            String original = file.getOriginalFilename();
            int idx = original.lastIndexOf('.');
            if (idx >= 0) ext = original.substring(idx);
        }

        String filename = id + ext;
        Path storagePath = Paths.get(uploadDir(), filename);
        String url = "/api/media/" + id;

        try {
            Files.write(storagePath, file.getBytes());
        } catch (IOException ex) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "파일을 저장할 수 없습니다.");
        }

        MediaAsset asset = new MediaAsset();
        asset.setId(id);
        asset.setUserId(user.userId());
        asset.setProjectId(projectUuid);
        asset.setOriginalName(file.getOriginalFilename());
        asset.setMimeType(file.getContentType());
        asset.setSize((int) file.getSize());
        asset.setStoragePath(storagePath.toString());
        asset.setUrl(url);
        return mediaRepo.save(asset);
    }

    public MediaAsset findById(UUID id) {
        return mediaRepo.findById(id).orElse(null);
    }

    public void assertOwnership(AiService.UserContext user, MediaAsset asset) {
        if (user == null || user.userId() == null) throw new ResponseStatusException(HttpStatus.FORBIDDEN, "로그인이 필요합니다.");
        if ("admin".equalsIgnoreCase(user.role())) return;
        if (!user.userId().equals(asset.getUserId())) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "권한이 없습니다.");
        }
    }

    public Object delete(AiService.UserContext user, UUID id) {
        MediaAsset asset = findById(id);
        if (asset == null) throw new ResponseStatusException(HttpStatus.NOT_FOUND, "파일을 찾을 수 없습니다.");
        assertOwnership(user, asset);

        mediaRepo.deleteById(id);
        try {
            Files.deleteIfExists(Path.of(asset.getStoragePath()));
        } catch (IOException ex) {
            // ignore
        }
        return java.util.Map.of("success", true);
    }
}
